benchmark/cxx-source/CxxVectorSum.swift

```diff
@@ -0,0 +1,89 @@
+//===--- CxxVectorSum.swift -----------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2012 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+// This is a benchmark that tracks how quickly Swift can sum up a C++ vector
+// as compared to the C++ implementation of such sum.
+
+import TestsUtils
+import CxxStdlibPerformance
+import Cxx
+
+public let benchmarks = [
+  BenchmarkInfo(
+      name: "CxxVectorOfU32SumInCxx",
+      runFunction: run_CxxVectorOfU32SumInCxx,
+      tags: [.validation, .bridging, .cxxInterop],
+      setUpFunction: create_CxxVectorOfU32),
+  BenchmarkInfo(
+    name: "CxxVectorOfU32SumInSwift_Fastest",
+    runFunction: run_CxxVectorOfU32SumInSwift_Fastest,
+    tags: [.validation, .bridging, .cxxInterop],
+    setUpFunction: create_CxxVectorOfU32),
+  BenchmarkInfo(
+    name: "CxxVectorOfU32SumInSwift",
+    runFunction: run_CxxVectorOfU32SumInSwift,
+    tags: [.validation, .bridging, .cxxInterop],
+    setUpFunction: create_CxxVectorOfU32)
+]
+
+// FIXME: compare CxxVectorOfU32SumInCxx to CxxVectorOfU32SumInSwift and
+// establish an expected threshold of performance, which when exceeded should
+// fail the benchmark.
+
+var vectorOfU32: VectorOfU32!
+
+func create_CxxVectorOfU32() {
+    vectorOfU32 = makeVector32(1_000_000)
+}
+
+@inline(never)
+public func run_CxxVectorOfU32SumInCxx(_ n: Int) {
+  // FIXME: Use no implicitly copyable, immutable borrow instead.
+  // https://github.com/apple/swift/issues/61454
+  let sum = testVector32Sum(&vectorOfU32)
+  blackHole(sum)
+}
+
+// This function should have comparable performance to `run_CxxVectorOfU32SumInCxx`.
+@inline(never)
+public func run_CxxVectorOfU32SumInSwift_Fastest(_ n: Int) {
+  var sum: UInt32 = 0
+  // begin mutating, end mutating needed to avoid copies right now.
+  var b = vectorOfU32.beginMutating()
```

Comment:
This pull the vector out of an optional, right? That's going to be a copy, I think. So it's not an entirely fair comparison.

Reply:
The mutable borrow doesn't pull it out. But possibly there's no way to avoid it with immutable borrow if we use the optional. Yeah let's avoid the optional thing for now.